swarm/swap_test.go

```diff
@@ -0,0 +1,468 @@
+				key, data, err := uploadRandomFileSize(sim.Service("swarm", id).(*Swarm), size)
+				if err != nil {
+					return err
+				}
+				log.Trace("file uploaded", "node", id, "key", key.String())
+				files = append(files, file{
+					addr:   key,
+					data:   data,
+					nodeID: id,
+				})
+			}
+		}
+
+		// File retrieval check is repeated until all uploaded files are retrieved from all nodes
+		// or until the timeout is reached.
+		for {
+			//we use a special retrieve function for swap which is optimized for parallel requests
+			//but does not leave many cascaded requests floating around
+			if retrieveForSwap(sim, files) == 0 {
+				break
+			}
+		}
+
+		for _, node := range sim.NodeIDs() {
+			item, ok := sim.NodeItem(node, bucketKeySwarm)
+			if !ok {
+				log.Error("No swarm")
+				return errors.New("no swarm")
+			}
+			swarm := item.(*Swarm)
+
+			subBalances := make(map[enode.ID]int64)
+
+			for _, n := range sim.NodeIDs() {
+				if node == n {
+					continue
+				}
+				balance, err := swarm.swap.GetPeerBalance(n)
+				if err == nil {
+					subBalances[n] = balance
+					log.Debug(fmt.Sprintf("Balance of node %s to node %s: %d", node.TerminalString(), n.TerminalString(), balance))
+				} else {
+					log.Debug(fmt.Sprintf("Node %s has no balance with node %s", node.TerminalString(), n.TerminalString()))
+				}
+			}
+			balancesMap[node] = subBalances
+		}
+
+		return nil
+
+	})
+
+	if result.Error != nil {
+		t.Fatal(result.Error)
+	}
+	if *printStats {
+		for k, v := range balancesMap {
+			fmt.Println(fmt.Sprintf("node %s balances:", k.TerminalString()))
+			for kk, vv := range v {
+				fmt.Println(fmt.Sprintf(".........with node %s: balance %d", kk.TerminalString(), vv))
+			}
+		}
+	}
+
+	/*
+		Assuming that in this case, balances should be symmetric too	I
+	*/
+
+	success := true
+	for k, mapForK := range balancesMap {
+		for n, balanceKwithN := range mapForK {
+			mapForSubK := balancesMap[n]
+			log.Trace(fmt.Sprintf("balance of %s with %s: %d", k.TerminalString(), n.TerminalString(), balanceKwithN))
+			log.Trace(fmt.Sprintf("balance of %s with %s: %d", n.TerminalString(), k.TerminalString(), mapForSubK[k]))
+			if balanceKwithN+mapForSubK[k] != 0 && balanceKwithN != 0 {
+				log.Error(fmt.Sprintf("Expected balances to be a+b = 0 AND balance(a) != 0, but they are not, balance k with n: %d, balance n with k: %d", balanceKwithN, mapForSubK[k]))
+				success = false
+			}
+		}
+	}
+
+	if !success {
+		t.Fatal("Expected balances to be symmetrical, but they were not")
+	}
+	log.Debug("test terminated")
+}
+
+// uploadFile, uploads a short file to the swarm instance
+// using the api.Put method.
+func uploadRandomFileSize(swarm *Swarm, size int) (storage.Address, string, error) {
+	b := make([]byte, size)
+	_, err := rand.Read(b)
+	if err != nil {
+		return nil, "", err
+	}
+	// uniqueness is very certain.
+	data := fmt.Sprintf("test content %s %x", time.Now().Round(0), b)
+	ctx := context.TODO()
```

Comment:
most of these comments are related to the original function which is `retrieve` in `network_test.go`, written by @janos. Originally I was going to reuse that function as-is, but because it is highly parallel, it resulted in problems for this test, so we created a custom version which does basically the same.

If these things need to be addressed, I request that this be done in a separate PR where we'd address these issues in a consistent manner for both tests.

Reply:
keep as is.